### PR TITLE
test(e2e): add test to verify user is logged out on its auth-token deletion

### PR DESCRIPTION
**Pull request by stasryzhov**
[ICU-8671](https://hashicorp.atlassian.net/browse/ICU-8671)

This PR creates new e2e test `TestUserIsLoggedOutWhenAuthTokenIsDeletedCli`. The test does the following: 
- Creates new account and assigns a new user to it
- Authenticates new user and assigns a name to its auth-token. The name is needed to be able to do operations via the user without having to authenticate it again (i.e. that would create a new auth-token)
- Verifies that the user is logged in (i.e. it can perform actions that require to be logged in)
- Deletes the user's auth-token by admin
- Verifies that the user is logged out (i.e. it cannot perform actions that require to be logged in)

Also, the PR refactors the `AuthenticateCli` function to take Options as a parameter. This allows to add additional args to the CLI command (e.g. `-token-name`)

[ICU-8671]: https://hashicorp.atlassian.net/browse/ICU-8671?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: testing/internal/e2e/boundary/authenticate.go [hunk 1]
```diff
@@ -53,11 +53,12 @@ func AuthenticateAdminCli(t testing.TB, ctx context.Context) {
 }
 
 // AuthenticateCli uses the cli to authenticate the specified Boundary instance
-func AuthenticateCli(t testing.TB, ctx context.Context, authMethodId string, loginName string, password string) {
+func AuthenticateCli(t testing.TB, ctx context.Context, authMethodId string, loginName string, password string, opt ...e2e.Option) {
```
Comment (stasryzhov) [inline testing/internal/e2e/boundary/authenticate.go:56]: Example of usage:
```go
// Authenticate user and assign a name to its auth token
boundary.AuthenticateCli(t, context.Background(), bc.AuthMethodId, testAccountName, acctPassword,
	e2e.WithArgs("-token-name", testAccountName),
)
```

File: testing/internal/e2e/tests/static/delete_auth_token_test.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+	// Check if user is logged out
+	output = e2e.RunCommand(ctx, "boundary",
+		e2e.WithArgs(
+			"auth-tokens", "list",
+			"-token-name", testAccountName,
+		),
+	)
+	require.Error(t, output.Err, fmt.Sprintf("User '%s' is still logged in", testAccountName))
+	t.Logf("User is logged out: %s", testAccountName)
```
Comment (moduli) [inline testing/internal/e2e/tests/static/delete_auth_token_test.go:80]: I wonder if there is a better way to validate this.

My original thought was to do something like this: https://github.com/hashicorp/boundary/blob/12aa7a0f5342d3569c9f189d73f4d7e54e5f9833/testing/internal/e2e/tests/static/session_end_delete_user_test.go#L61-L81 (use `GetAuthenticationTokenCli` to get you a token - I assume this creates an auth token that appears in `boundary auth-tokens list`). The admin would then delete that auth token, and then the test would try to use the token again with some other command, and that should fail.

**Review by moduli — CHANGES_REQUESTED** [review]

File: testing/internal/e2e/tests/static/delete_auth_token_test.go (hunk 2 above)
Comment (moduli) [reply testing/internal/e2e/tests/static/delete_auth_token_test.go:80]: After discussion with Stan, the current is fine.

File: testing/internal/e2e/boundary/authtoken.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+	"encoding/json"
+	"fmt"
+	"testing"
+
+	"github.com/hashicorp/boundary/api/authtokens"
+	"github.com/hashicorp/boundary/testing/internal/e2e"
+	"github.com/stretchr/testify/require"
+)
+
+// GetAuthenticationTokenIdCli uses the CLI to get an auth-token ID by its name
```
Comment (moduli) [inline testing/internal/e2e/boundary/authtoken.go:17]: I think I was confused by this comment. To me, this reads as if you're looking up an auth-token ID using its name, but based on our earlier conversation, this is using the token name to authenticate the command.

Two things:
- How do we know we're retrieving the desired auth token here? The test is assuming this is the user's auth token? Is that the only one that is in the list?
- To keep things methods in this package consistent, should we not make `tokenName` required?

Comment (stasryzhov) [reply testing/internal/e2e/boundary/authtoken.go]: After discussion with @moduli, we decided to keep current approach and rename the method for more clarity.
Also, I've updated the comment to provide more detailed description of what the method does

File: testing/internal/e2e/tests/static/auth_token_delete_test.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+	ctx := context.Background()
+	boundary.AuthenticateAdminCli(t, ctx)
+	newAccountId, acctPassword := boundary.CreateNewAccountCli(t, ctx, bc.AuthMethodId, testAccountName)
+	t.Cleanup(func() {
+		boundary.AuthenticateAdminCli(t, context.Background())
+		output := e2e.RunCommand(ctx, "boundary",
+			e2e.WithArgs("accounts", "delete", "-id", newAccountId),
+		)
+		require.NoError(t, output.Err, string(output.Stderr))
+		t.Logf("Successfully deleted account '%s'", newAccountId)
```
Comment (moduli) [inline testing/internal/e2e/tests/static/auth_token_delete_test.go:38]: Suggest to remove this log as it's in the cleanup and not the actual test

File: testing/internal/e2e/tests/static/auth_token_delete_test.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+	})
+	newUserId := boundary.CreateNewUserCli(t, ctx, "global")
+	t.Cleanup(func() {
+		boundary.AuthenticateAdminCli(t, context.Background())
+		output := e2e.RunCommand(ctx, "boundary",
+			e2e.WithArgs("users", "delete", "-id", newUserId),
+		)
+		require.NoError(t, output.Err, string(output.Stderr))
+		t.Logf("Successfully deleted user '%s'", newUserId)
```
Comment (moduli) [inline testing/internal/e2e/tests/static/auth_token_delete_test.go:47]: Suggest to remove this log as it's in the cleanup and not the actual test

File: testing/internal/e2e/tests/static/auth_token_delete_test.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+	output = e2e.RunCommand(ctx, "boundary",
+		e2e.WithArgs(
+			"auth-tokens", "list",
+			"-token-name", testAccountName,
+		),
+	)
+
+	// Expect error as the user got logged out
+	require.Error(t, output.Err, fmt.Sprintf("User '%s' is still logged in", testAccountName))
+	t.Logf("User is logged out: %s", testAccountName)
```
Comment (moduli) [inline testing/internal/e2e/tests/static/auth_token_delete_test.go:88]:
```suggestion
	t.Logf("Successfully verified that token name is invalid")
```

**Review by moduli — APPROVED** [review]
Some additional minor comments about logging for consistency purposes.

**Pull request merged**